systemd/cloud-init-generator.tmpl

```diff
@@ -40,105 +42,47 @@ debug() {
     echo "$@" >> "$LOG"
 }
 
-etc_file() {
-    local pprefix="${1:-/etc/cloud/cloud-init.}"
-    _RET="unset"
-    [ -f "${pprefix}$ENABLE" ] && _RET="$ENABLE" && return 0
-    [ -f "${pprefix}$DISABLE" ] && _RET="$DISABLE" && return 0
-    return 0
-}
-
-read_proc_cmdline() {
-    # return /proc/cmdline for non-container, and /proc/1/cmdline for container
-    local ctname="systemd"
-    if [ -n "$CONTAINER" ] && ctname=$CONTAINER ||
-        systemd-detect-virt --container --quiet; then
-        if { _RET=$(tr '\0' ' ' < /proc/1/cmdline); } 2>/dev/null; then
-            _RET_MSG="container[$ctname]: pid 1 cmdline"
-            return
-        fi
-        _RET=""
-        _RET_MSG="container[$ctname]: pid 1 cmdline not available"
-        return 0
-    fi
-
-    _RET_MSG="/proc/cmdline"
-    read _RET < /proc/cmdline
-}
-
-kernel_cmdline() {
-    local cmdline="" tok=""
-    if [ -n "${KERNEL_CMDLINE+x}" ]; then
-        # use KERNEL_CMDLINE if present in environment even if empty
-        cmdline=${KERNEL_CMDLINE}
-        debug 1 "kernel command line from env KERNEL_CMDLINE: $cmdline"
-    else
-        read_proc_cmdline && cmdline="$_RET" &&
-                debug 1 "kernel command line ($_RET_MSG): $cmdline"
-    fi
-    _RET="unset"
-    cmdline=" $cmdline "
-    tok=${cmdline##* cloud-init=}
-    [ "$tok" = "$cmdline" ] && _RET="unset"
-    tok=${tok%% *}
-    [ "$tok" = "$ENABLE" -o "$tok" = "$DISABLE" ] && _RET="$tok"
-    return 0
-}
-
-default() {
-    _RET="$ENABLE"
-}
-
-check_for_datasource() {
-    local ds_rc=""
-    if [ ! -x "$dsidentify" ]; then
-        debug 1 "no ds-identify in $dsidentify"
-        return 0
-    fi
-    $dsidentify
-    ds_rc=$?
-    debug 1 "ds-identify rc=$ds_rc"
-    if [ "$ds_rc" = "0" ]; then
-        return 0
-    fi
-    return 1
-}
-
 main() {
     local normal_d="$1" early_d="$2" late_d="$3"
     local target_name="multi-user.target" gen_d="$early_d"
     local link_path="$gen_d/${target_name}.wants/${CLOUD_TARGET_NAME}"
-    local ds=""
+    local ds="" ret=""
 
     debug 1 "$0 normal=$normal_d early=$early_d late=$late_d"
     debug 2 "$0 $*"
 
-    local search result="error" ret=""
-    for search in kernel_cmdline etc_file default; do
-        if $search; then
-            debug 1 "$search found $_RET"
-            [ "$_RET" = "$ENABLE" -o "$_RET" = "$DISABLE" ] &&
-                result=$_RET && break
+    # ds=found => enable
+    # ds=notfound => disable
+    # <any> => disable
+    debug 1 "checking for datasource"
+
+    if [ ! -x "$dsidentify" ]; then
+        debug 1 "no ds-identify in $dsidentify"
+        ds=0
+    fi
+    $dsidentify
+    ds=$?
+    debug 1 "ds-identify rc=$ds"
+
+    if [ "$ds" = "1" ]; then
+        debug 1 "cloud-init is enabled but no datasource found, disabling"
```

Comment:
If a cloud-init is installed but disabled, the misleading message `cloud-init is enabled but no datasource found, disabling` will now be printed by the generator on boot, and ds-identify will output misleading messages attempting to detect clouds. 

ds-identify will likely write incorrect /run/cloud-init/cloud.cfg configuration as if it was enabling datasources. 

$RUN_ENABLED_FILE` (`/run/cloud-init/cloud-init.enabled`/`.disabled`) will also be set to the wrong state. I'm not sure if this file is used for anything since dhclient hooks were removed in #2051, but it could be used as an indicator for other external scripts. cloud-init.enabled also doesn't appear to be created when not using the generator.

Reply:
The message will only be printed if no datasource is found. This is still misleading, no question, but it is not unconditional. In the case of a  "happy path", i.e. dsidentify runs and finds a data source the message will not be printed.

That said if cloud-init is disabled, why run dsitentify at all?

Reply:
Even though this text in /run/cloud-init/cloud-init-generator.log is not machine readable, and I don't think we have scripts parsing and relying on the output of this file, I'd like us to specialize that message when cloud-init is disabled. Then we better represent  the actual state during generator time-frame when we see cloud-init disabled and don't have to fallback to querying cloud-init status or systemctl status cloud-init.target etc to get that info.